Bootstrap from more permanent URLs

SimonSapin opened the pull request:
The `rust-lang-ci` S3 bucket is ephemeral. `static-rust-lang-org.s3.amazonaws.com` is not going away soon, but using `static.rust-lang.org` when possible keeps things working if it ever does.

https://internals.rust-lang.org/t/updates-on-rusts-ci-uploads/6062
https://internals.rust-lang.org/t/public-stable-rust-services/6072

We’ll still need to find a solution for "alt" rustc builds. In the meantime, this is a step.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/18942)
<!-- Reviewable:end -->

SimonSapin committed:
Download "non-alt" rustc builds from more permanent URLs

The `rust-lang-ci` S3 bucket is ephemeral:

https://internals.rust-lang.org/t/updates-on-rusts-ci-uploads/6062
https://internals.rust-lang.org/t/public-stable-rust-services/6072

We’ll still need to find a solution for "alt" rustc builds.
In the meantime, this is a step.

mbrubeck commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit efdaa0d has been approved by `mbrubeck`

<!-- @bors-servo r=mbrubeck efdaa0d00ac286b4dc5ad8d31e68c98a079b8917 -->

SimonSapin committed:
Bootstrap: print URLs being downloaded.

SimonSapin committed:
Bootstrap from official static.rust-lang.org when SNI is available

mbrubeck commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 664f282 has been approved by `mbrubeck`

<!-- @bors-servo r=mbrubeck 664f2829a9d7e8f5a158111b80e71eb8b830f283 -->

SimonSapin committed:
Use servo-deps AWS S3 bucket instead of rust-lang-ci for OpenSSL

mbrubeck commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit a931d3b has been approved by `mbrubeck`

<!-- @bors-servo r=mbrubeck a931d3b840d25bf536eeae5e457b02f9e338feaa -->

bors-servo commented:
:hourglass: Testing commit a931d3b840d25bf536eeae5e457b02f9e338feaa with merge 3beb62bf21d2e4403ce66548a2a30796a0e1e42a...

bors-servo commented:
:broken_heart: Test failed - [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/5100)

SimonSapin commented:
```
URLError: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:590)>

  File "c:\buildbot\slave\windows-msvc-dev\build\python\servo\build_commands.py", line 233, in build
    self.ensure_bootstrapped(target=target)
  File "c:\buildbot\slave\windows-msvc-dev\build\python\servo\command_base.py", line 618, in ensure_bootstrapped
    stable=self._use_stable_rust)
  File "c:\buildbot\slave\windows-msvc-dev\build\python\_virtualenv\Lib\site-packages\mach\registrar.py", line 123, in dispatch
    return self._run_command_handler(handler, context=context, **kwargs)
  File "c:\buildbot\slave\windows-msvc-dev\build\python\_virtualenv\Lib\site-packages\mach\registrar.py", line 90, in _run_command_handler
    result = fn(**kwargs)
  File "c:\buildbot\slave\windows-msvc-dev\build\python\servo\bootstrap_commands.py", line 97, in bootstrap_rustc
    nightly_commit_hash = toml.load(urllib2.urlopen(channel))["pkg"]["rustc"]["git_commit_hash"]
  File "C:\Python27\lib\urllib2.py", line 154, in urlopen
    return opener.open(url, data, timeout)
  File "C:\Python27\lib\urllib2.py", line 431, in open
    response = self._open(req, data)
  File "C:\Python27\lib\urllib2.py", line 449, in _open
    '_open', req)
  File "C:\Python27\lib\urllib2.py", line 409, in _call_chain
    result = func(*args)
  File "C:\Python27\lib\urllib2.py", line 1240, in https_open
    context=self._context)
  File "C:\Python27\lib\urllib2.py", line 1197, in do_open
    raise URLError(err)
```

Retrying in case this an intermittent network error, though it could also be that Python on this Windows builder is not configure with the appropriate root certificates to verify `static.rust-lang.org`.

@bors-servo retry

SimonSapin commented:
@bors-servo r-

SimonSapin commented:
Based on messing around with salt:

* Our Mac builders have Python 2.7.13 with `ssl.HAS_SNI == True`, and connecting to `static.rust-lang.org` works.
* Our Linux builders have Python 2.7.6 without a `ssl.HAS_SNI` attribute, so this PR uses the fallback hostname which works.
* Our Windows builders have Python 2.7.12 with `ssl.HAS_SNI == True`, and fail to verify the certificate when connecting to `static.rust-lang.org`. This may or may not be related to SNI.

SimonSapin commented:
> Our Windows builders have Python 2.7.12 with `ssl.HAS_SNI == True`, and fail to verify the certificate when connecting to `static.rust-lang.org`. This may or may not be related to SNI.

These Windows builders *can* successfully connect to https://static-rust-lang-org.s3.amazonaws.com/

SimonSapin committed:
Bootstrap: use 'certifi' package to obtain up-to-date root CA list.

The default root CA store on our Windows CI builders does not
trust static.rust-lang.org

jdm commented:
@bors-servo: r=jdm,mbrubeck

bors-servo commented:
:pushpin: Commit aa62942 has been approved by `jdm,mbrubeck`

<!-- @bors-servo r=jdm,mbrubeck aa62942fbd836279da8ea41d2e5d4ca9dd61fad3 -->

bors-servo commented:
:hourglass: Testing commit aa62942fbd836279da8ea41d2e5d4ca9dd61fad3 with merge d4c63d835bee38b6afb03c83a53b7de1963c87a5...

bors-servo commented:
:broken_heart: Test failed - [android](http://build.servo.org/builders/android/builds/9451)

jdm commented:
`TypeError: urlopen() got an unexpected keyword argument 'cafile'`

SimonSapin commented:
The `cafile` parameter of `urllib2.urlopen` is new in 2.7.9…

SimonSapin committed:
Only pass cafile argument to urlopen in Python versions that support it.

jdm commented:
@bors-servo: r=jdm,mbrubeck
*yakkety sax plays in the distance*

bors-servo commented:
:pushpin: Commit 404c722 has been approved by `jdm,mbrubeck`

<!-- @bors-servo r=jdm,mbrubeck 404c722920043caf548bf10c0242aac5d22a96b3 -->

bors-servo commented:
:hourglass: Testing commit 404c722920043caf548bf10c0242aac5d22a96b3 with merge ce5488ee74d8e418f52db99e60dffe236bf90fb2...

bors-servo commented:
:broken_heart: Test failed - [arm32](http://build.servo.org/builders/arm32/builds/9683)

jdm commented:
@bors-servo: retry
* maybe the virtualenv is busted?

bors-servo commented:
:hourglass: Testing commit 404c722920043caf548bf10c0242aac5d22a96b3 with merge 5e87ed418080fd54b1a8e6cbd0874901c9f525ea...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/6560)

mbrubeck commented:
<details>

```
Error running mach:

    ['test-css', '--release', '--processes', '16', '--log-raw', 'test-css.log', '--log-errorsummary', 'css-errorsummary.log', '--always-succeed']

The error occurred in code that was called by the mach command. This is either
a bug in the called code itself or in the way that mach is calling it.

You should consider filing a bug for this issue.

If filing a bug, please include the full output of mach, including this error
message.

The details of the failure are as follows:

AttributeError: 'module' object has no attribute 'QueuedProxyLogger'

  File "/home/servo/buildbot/slave/linux-rel-css/build/python/servo/testing_commands.py", line 618, in test_css
    ret = self.run_test_list_or_dispatch(kwargs["test_list"], "css", self._test_css, **kwargs)
  File "/home/servo/buildbot/slave/linux-rel-css/build/python/servo/testing_commands.py", line 443, in run_test_list_or_dispatch
    return correct_function(**kwargs)
  File "/home/servo/buildbot/slave/linux-rel-css/build/python/servo/testing_commands.py", line 626, in _test_css
    return self.wptrunner(run_file, **kwargs)
  File "/home/servo/buildbot/slave/linux-rel-css/build/python/servo/testing_commands.py", line 484, in wptrunner
    return run_globals["run_tests"](**kwargs)
  File "/home/servo/buildbot/slave/linux-rel-css/build/tests/wpt/run_css.py", line 13, in run_tests
    return run.run_tests(paths=paths, **kwargs)
  File "/home/servo/buildbot/slave/linux-rel-css/build/tests/wpt/run.py", line 46, in run_tests
    success = wptrunner.run_tests(**kwargs)
  File "/home/servo/buildbot/slave/linux-rel-css/build/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/wptrunner.py", line 181, in run_tests
    env_extras) as test_environment:
  File "/home/servo/buildbot/slave/linux-rel-css/build/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/environment.py", line 104, in __enter__
    self.setup_server_logging()
  File "/home/servo/buildbot/slave/linux-rel-css/build/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/environment.py", line 171, in setup_server_logging
    server_logger = proxy.QueuedProxyLogger(server_logger)
program finished with exit code 1
```

</details>

jdm commented:
@bors-servo: retry
* #18961

bors-servo commented:
:hourglass: Testing commit 404c722920043caf548bf10c0242aac5d22a96b3 with merge 041bd626ace013f93fa7fe101c70f36543fc9b0d...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/9477), [arm32](http://build.servo.org/builders/arm32/builds/9706), [arm64](http://build.servo.org/builders/arm64/builds/9739), [linux-dev](http://build.servo.org/builders/linux-dev/builds/9757), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/6579), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/6610), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/9827), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/2250), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/2265), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/6523), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/6578), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/2319), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/2293), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/5133)
Approved by: jdm,mbrubeck
Pushing 041bd626ace013f93fa7fe101c70f36543fc9b0d to master...
